### PR TITLE
dataone now uses XMLHarvester base class

### PR DESCRIPTION
**Pull request by fabianvf**

### CODE REVIEW

**Comment by fabianvf** [comment]
@chrisseto @erinspace

File: scrapi/harvesters/dataone.py [hunk 1]
```diff
@@ -20,18 +20,135 @@
...
+def process_doi(service_id, doc_doi):
+    doi_re = '10\\.\\d{4}/\\w*\\.\\w*(/\\w*)?'
+    try:
+        regexed_doi = re.search(doi_re, service_id).group(0)
+        doi = regexed_doi
+    except AttributeError:
+        try:
+            doc_doi = doc_doi.replace('doi:', '')
+            regexed_doi = re.search(doi_re, doc_doi).group(0)
+        except (IndexError, AttributeError):
```
Comment (erinspace) [inline scrapi/harvesters/dataone.py]: When would an IndexError be raised?

File: scrapi/harvesters/dataone.py [hunk 2]
```diff
@@ -20,18 +20,135 @@
...
+            'doi': ("str[@name='id']/node()", "arr[@name='isDocumentedBy']/str/node()", process_doi),
+            'url': ("str[@name='id']/node()", "//str[@name='dataUrl']/node()", lambda x, y: y if 'http' in y else x if 'http' in x else '')
+        },
+        'tags': ("//arr[@name='keywords']/str/node()", lambda x: x if isinstance(x, list) else [x]),
+        'dateUpdated': ("str[@name='dateModified']/node()", lambda x: unicode(parse(x).isoformat())),
+        'title': "str[@name='title']/node()",
+        'description': "str[@name='abstract']/node()"
+    }
 
     DEFAULT_ENCODING = 'UTF-8'
```
Comment (erinspace) [inline scrapi/harvesters/dataone.py]: I don't think we need this constant anymore

Comment (erinspace) [inline scrapi/harvesters/dataone.py]: Or rather, we as a group should re-take a look at the copy to unicode function! This is OK as is - my bad!

File: scrapi/harvesters/dataone.py [hunk 3]
```diff
@@ -20,18 +20,135 @@
...
     DATAONE_SOLR_ENDPOINT = 'https://cn.dataone.org/cn/v1/query/solr/'
```
Comment (chrisseto) [inline scrapi/harvesters/dataone.py]: Can this be moved to the top of the file?

File: scrapi/harvesters/dataone.py [hunk 4]
```diff
@@ -20,18 +20,135 @@
...
+            'isDocumentedBy': "arr[@name='isDocumentedBy']/str/node()",
+        },
+        'contributors': ("str[@name='author']/node()", "str[@name='submitter']/node()", "arr[@name='origin']/str/node()", process_contributors),
+        'id': {
+            'serviceID': "str[@name='id']/node()",
+            'doi': ("str[@name='id']/node()", "arr[@name='isDocumentedBy']/str/node()", process_doi),
+            'url': ("str[@name='id']/node()", "//str[@name='dataUrl']/node()", lambda x, y: y if 'http' in y else x if 'http' in x else '')
+        },
+        'tags': ("//arr[@name='keywords']/str/node()", lambda x: x if isinstance(x, list) else [x]),
+        'dateUpdated': ("str[@name='dateModified']/node()", lambda x: unicode(parse(x).isoformat())),
```
Comment (chrisseto) [inline scrapi/harvesters/dataone.py]: Use `.decode(utf-8)` over unicode.

**Comment by fabianvf** [comment]
@chrisseto @erinspace ready for pass 2

**Comment by erinspace** [comment]
Looks good to me! :+1:

**Comment by fabianvf** [comment]
@chrisseto :shipit:

**Comment by fabianvf** [comment]
@chrisseto plz :help: :shipit:

**Pull request merged**